scheduler/src/cook/mesos/sandbox.clj

```diff
@@ -158,78 +151,93 @@
-             (counter-reset sandbox-updater-unprocessed-count (count result-map))
-             result-map))))
+  "Aggregates the unprocessed task-id into the unprocessed-host->task-ids-atom for a given host."
+  [unprocessed-host->task-ids-atom host task-id]
+  (swap! unprocessed-host->task-ids-atom
+         (fn [unprocessed-host->task-ids-in-atom]
+           (let [unprocessed-task-ids (get unprocessed-host->task-ids-in-atom host #{})]
+             (if-not (contains? unprocessed-task-ids task-id)
+               (do
+                 (counters/inc! sandbox-updater-unprocessed-count)
+                 (->> task-id
+                      (conj unprocessed-task-ids)
+                      (assoc unprocessed-host->task-ids-in-atom host)))
+               unprocessed-host->task-ids-in-atom))))
+  (histograms/update! sandbox-updater-unprocessed-entries (counters/value sandbox-updater-unprocessed-count)))
 
 (defn- remove-processed-task-ids!
-  "Removes the processed-task-ids from the unprocessed-task-id->sandbox-atom."
-  [unprocessed-task-id->sandbox-atom processed-task-ids]
-  (swap! unprocessed-task-id->sandbox-atom
-         (fn [unprocessed-task-id->sandbox-in-atom]
-           (let [result-map (apply dissoc unprocessed-task-id->sandbox-in-atom processed-task-ids)]
-             (counter-reset sandbox-updater-unprocessed-count (count result-map))
-             result-map))))
+  "Removes the processed-task-ids from the unprocessed-host->task-ids-atom for a given host."
+  [unprocessed-host->task-ids-atom host task-ids]
+  (swap! unprocessed-host->task-ids-atom
+         (fn [unprocessed-host->task-ids-in-atom]
+           (let [unprocessed-task-ids (get unprocessed-host->task-ids-in-atom host #{})
+                 unprocessed-task-ids' (set/difference unprocessed-task-ids task-ids)]
+             (if (not= unprocessed-task-ids unprocessed-task-ids')
+               (do
+                 (->> (- (count unprocessed-task-ids') (count unprocessed-task-ids))
+                      (counters/inc! sandbox-updater-unprocessed-count))
+                 (if (seq unprocessed-task-ids')
+                   (assoc unprocessed-host->task-ids-in-atom host unprocessed-task-ids')
+                   (dissoc unprocessed-host->task-ids-in-atom host)))
+               unprocessed-host->task-ids-in-atom))))
+  (histograms/update! sandbox-updater-unprocessed-entries (counters/value sandbox-updater-unprocessed-count)))
+
+(defn- process-task-id->sandbox-directory-on-host
+  "Processes the result of the indexed task-id->sandbox-directory on a host.
+   First, the tasks which have pending sandbox directory requests are filtered based on existing requests.
+   The filtered version of tasks without sandbox directories is then synced into the task-id->sandbox-agent."
+  [{:keys [task-id->sandbox-agent unprocessed-host->task-ids-atom]} host unprocessed-task-ids-prev task-id->sandbox-directory]
```

Comment:
What is the `unprocessed-task-ids-prev` for? I don't think I understand the logic here.

Reply:
I updated `refresh-agent-cache-entry` to further handle this possibility by ensuring the sync of the host is triggered when this scnerio occurs.